integrationTests/chainSimulator/staking/delegation_test.go

```diff
@@ -338,23 +347,470 @@ func testBLSKeyIsInAuction(
+	require.Equal(t, "1250000000000000000000", big.NewInt(0).SetBytes(output.ReturnData[0]).String())
+	require.Equal(t, zeroValue, getBLSTopUpValue(t, metachainNode, delegationContractAddressBytes))
+
+	output, err = executeQuery(cs, core.MetachainShardId, delegationContractAddressBytes, "getUserActiveStake", [][]byte{delegator2Bytes})
+	require.Nil(t, err)
+	require.Equal(t, "0", big.NewInt(0).SetBytes(output.ReturnData[0]).String())
+
+	// still staked until epoch change
+	output, err = executeQuery(cs, core.MetachainShardId, delegationContractAddressBytes, "getAllNodeStates", nil)
+	require.Nil(t, err)
+	stakedKeys, notStakedKeys, unStakedKeys = getNodesFromContract(output.ReturnData)
+	require.Equal(t, 1, len(stakedKeys))
+	require.Equal(t, blsKeys[0], hex.EncodeToString(stakedKeys[0]))
+	require.Equal(t, 0, len(notStakedKeys))
+	require.Equal(t, 0, len(unStakedKeys))
+
+	err = cs.GenerateBlocksUntilEpochIsReached(targetEpoch + 1)
+	require.Nil(t, err)
+
+	output, err = executeQuery(cs, core.MetachainShardId, delegationContractAddressBytes, "getAllNodeStates", nil)
+	require.Nil(t, err)
+	stakedKeys, notStakedKeys, unStakedKeys = getNodesFromContract(output.ReturnData)
+	require.Equal(t, 0, len(stakedKeys))
+	require.Equal(t, 0, len(notStakedKeys))
+	require.Equal(t, 1, len(unStakedKeys))
+	require.Equal(t, blsKeys[0], hex.EncodeToString(unStakedKeys[0]))
+}
+
+func generateWalletAddressBytes() []byte {
+	buff := make([]byte, walletAddressBytesLen)
+	_, _ = rand.Read(buff)
+
+	return buff
+}
+
+func executeQuery(cs chainSimulatorIntegrationTests.ChainSimulator, shardID uint32, scAddress []byte, funcName string, args [][]byte) (*dataVm.VMOutputApi, error) {
+	output, _, err := cs.GetNodeHandler(shardID).GetFacadeHandler().ExecuteSCQuery(&process.SCQuery{
+		ScAddress: scAddress,
+		FuncName:  funcName,
+		Arguments: args,
+	})
+	return output, err
+}
+
+func addNodesTxData(blsKeys []string, sigs [][]byte) string {
+	txData := "addNodes"
+
+	for i := range blsKeys {
+		txData = txData + "@" + blsKeys[i] + "@" + hex.EncodeToString(sigs[i])
+	}
+
+	return txData
+}
+
+func getSignatures(msg []byte, blsKeys [][]byte) [][]byte {
+	signer := mclsig.NewBlsSigner()
+
+	signatures := make([][]byte, len(blsKeys))
+	for i, blsKey := range blsKeys {
+		sk, _ := signing.NewKeyGenerator(mcl.NewSuiteBLS12()).PrivateKeyFromByteArray(blsKey)
+		signatures[i], _ = signer.Sign(sk, msg)
+	}
+
+	return signatures
+}
+
+func getNodesFromContract(returnData [][]byte) ([][]byte, [][]byte, [][]byte) {
+	var stakedKeys, notStakedKeys, unStakedKeys [][]byte
+
+	// Placeholder for the current list being populated
+	var currentList *[][]byte
+
+	for _, data := range returnData {
+		switch string(data) {
+		case "staked":
+			currentList = &stakedKeys
+		case "notStaked":
+			currentList = &notStakedKeys
+		case "unStaked":
+			currentList = &unStakedKeys
+		default:
+			if currentList != nil {
+				*currentList = append(*currentList, data)
+			}
+		}
+	}
```

Comment:
```   
      var stakedKeys, notStakedKeys, unStakedKeys [][]byte
 
      for i := 0; i < len(returnData); i += 2 {
          switch string(returnData[i]) {
          case "staked":
              stakedKeys = append(stakedKeys, returnData[i+1])
          case "notStaked":
              notStakedKeys = append(notStakedKeys, returnData[i+1])
          case "unStaked":
              unStakedKeys = append(unStakedKeys, returnData[i+1])
          }
      }
 
      return stakedKeys, notStakedKeys, unStakedKeys
```
i think something like this might be easier to understand for this loop

Reply:
👍